File: ts/tests/trust.test.ts
import { describe, it, expect } from "vitest";
import {
  DefaultTrustModel,
  TrustMetrics,
  defaultTrustConfig,
} from "../src/trust.js";
import { Actor, ActorType, ActorStatus, InMemoryActorStore } from "../src/actor.js";
import {
  ActorId,
  ConversationId,
  DomainScope,
  EventId,
  EventType,
  Hash,
  NonEmpty,
  Score,
  Signature,
  PublicKey,
  Weight,
} from "../src/types.js";
import { Event, newEventId } from "../src/event.js";

// ── Helpers ──────────────────────────────────────────────────────────────

function testPublicKey(b: number): PublicKey {
  const key = new Uint8Array(32);
  key[0] = b;
  return new PublicKey(key);
}

function testActor(name: string, b: number): Actor {
  const store = new InMemoryActorStore();
  return store.register(testPublicKey(b), name, ActorType.Human);
}

function testTrustEvent(actorId: ActorId, prev: number, curr: number, domain = "general"): Event {
  const content: Record<string, unknown> = {
    actor: actorId.value,
    previous: prev,
    current: curr,
    domain: domain,
    cause: "019462a0-0000-7000-8000-000000000001",
  };
  const causeId = new EventId("019462a0-0000-7000-8000-000000000001");
  return new Event(
    1,
    newEventId(),
    new EventType("trust.updated"),
    Date.now() * 1_000_000,
    actorId,
    content,
    NonEmpty.of([causeId]),
    new ConversationId("conv_test000000000000000000000001"),
    Hash.zero(),
    Hash.zero(),
    new Signature(new Uint8Array(64)),
  );
}

function testNonTrustEvent(actorId: ActorId): Event {
  const content: Record<string, unknown> = {
    message: "hello",
  };
  const causeId = new EventId("019462a0-0000-7000-8000-000000000001");
  return new Event(
    1,
    newEventId(),
    new EventType("system.bootstrapped"),
    Date.now() * 1_000_000,
    actorId,
    content,
    NonEmpty.of([causeId]),
    new ConversationId("conv_test000000000000000000000001"),
    Hash.zero(),
    Hash.zero(),
    new Signature(new Uint8Array(64)),
  );
}

// ── Tests ────────────────────────────────────────────────────────────────

describe("DefaultTrustModel", () => {
  it("initialScore", () => {
    const model = new DefaultTrustModel();
    const a = testActor("Alice", 1);

    const metrics = model.score(a);
    expect(metrics.overall.value).toBe(0.0);
    expect(metrics.confidence.value).toBe(0.0);
  });

  it("updateIncreasesTrust", () => {
    const model = new DefaultTrustModel();
    const a = testActor("Alice", 1);

    const ev = testTrustEvent(a.id, 0.0, 0.05);
    const metrics = model.update(a, ev);
    expect(metrics.overall.value).toBeGreaterThan(0.0);
  });

  it("updateDecreasesTrust", () => {
    const model = new DefaultTrustModel();
    const a = testActor("Alice", 1);

    // First increase
    const ev1 = testTrustEvent(a.id, 0.0, 0.08);
    model.update(a, ev1);

    // Then decrease
    const ev2 = testTrustEvent(a.id, 0.08, 0.0);
    const metrics = model.update(a, ev2);
    expect(metrics.overall.value).toBeLessThan(0.08);
  });

  it("updateClampedToMaxAdjustment", () => {
    const model = new DefaultTrustModel();
    const a = testActor("Alice", 1);

    // Try to increase by 0.5 — should be clamped to MaxAdjustment (0.1)
    const ev = testTrustEvent(a.id, 0.0, 0.5);
    const metrics = model.update(a, ev);
    expect(metrics.overall.value).toBeLessThanOrEqual(0.1);
  });

  it("updateDeduplication", () => {
    const model = new DefaultTrustModel();
    const a = testActor("Alice", 1);

    const ev = testTrustEvent(a.id, 0.0, 0.05);
    const m1 = model.update(a, ev);

    // Same event again — should not change score
    const m2 = model.update(a, ev);
    expect(m2.overall.value).toBe(m1.overall.value);
    expect(m2.evidence.length).toBe(m1.evidence.length);
  });

  it("updateTrendPositive", () => {
    const model = new DefaultTrustModel();
    const a = testActor("Alice", 1);

    const ev = testTrustEvent(a.id, 0.0, 0.05);
    const metrics = model.update(a, ev);
    expect(metrics.trend.value).toBeGreaterThan(0);
  });

  it("updateTrendNegative", () => {
    const model = new DefaultTrustModel();
    const a = testActor("Alice", 1);

    // First increase to have something to decrease from
    const ev1 = testTrustEvent(a.id, 0.0, 0.05);
    model.update(a, ev1);

    // Then decrease — trend should go down
    const ev2 = testTrustEvent(a.id, 0.1, 0.0);
    const metrics = model.update(a, ev2);
    // After +0.1 then -0.1, trend should be 0
    expect(metrics.trend.value).toBeLessThanOrEqual(0.1);
  });

  it("scoreInDomain", () => {
    const model = new DefaultTrustModel();
    const a = testActor("Alice", 1);

    // Update with domain-specific evidence
    const ev = testTrustEvent(a.id, 0.0, 0.05, "code_review");
    model.update(a, ev);

    const domain = new DomainScope("code_review");
    const metrics = model.scoreInDomain(a, domain);
    expect(metrics.overall.value).toBeGreaterThan(0.0);
  });

  it("scoreInDomainFallback", () => {
    const model = new DefaultTrustModel();
    const a = testActor("Alice", 1);

    // No domain-specific evidence, should fall back to global with halved confidence
    const domain = new DomainScope("code_review");
    const metrics = model.scoreInDomain(a, domain);
    expect(metrics.overall.value).toBe(0.0);
    expect(metrics.confidence.value).toBe(0.0); // 0 evidence * 0.5 = 0
  });

  it("scoreInDomainFallbackWithEvidence", () => {
    const model = new DefaultTrustModel();
    const a = testActor("Alice", 1);

    // Add non-trust event to build evidence (no domain info)
    const ev = testNonTrustEvent(a.id);
    model.update(a, ev);

    // Query a domain that has no specific data
    const domain = new DomainScope("unknown_domain");
    const metrics = model.scoreInDomain(a, domain);
    // Confidence should be halved compared to global
    const globalMetrics = model.score(a);
    expect(metrics.confidence.value).toBeCloseTo(globalMetrics.confidence.value * 0.5, 10);
  });

  it("decay", () => {
    const model = new DefaultTrustModel();
    const a = testActor("Alice", 1);

    // Build up some trust
    for (let i = 0; i < 5; i++) {
      const ev = testTrustEvent(a.id, 0.0, 0.1);
      model.update(a, ev);
    }

    const before = model.score(a);

    // Decay 10 days (in seconds)
    model.decay(a, 10 * 86400);

    const after = model.score(a);
    expect(after.overall.value).toBeLessThan(before.overall.value);
  });

  it("decayNegativeDuration", () => {
    const model = new DefaultTrustModel();
    const a = testActor("Alice", 1);

    // Build some trust
    const ev = testTrustEvent(a.id, 0.0, 0.05);
    model.update(a, ev);

    const before = model.score(a);

    // Negative duration — should not change score
    const metrics = model.decay(a, -100);
    expect(metrics.overall.value).toBe(before.overall.value);
  });

  it("updateBetween", () => {
    const model = new DefaultTrustModel();
    const from = testActor("Alice", 1);
    const to = testActor("Bob", 2);

    const ev = testTrustEvent(to.id, 0.0, 0.05);
    const metrics = model.updateBetween(from, to, ev);
    expect(metrics.overall.value).toBeGreaterThan(0.0);
    expect(metrics.actor.value).toBe(to.id.value);
  });

  it("betweenNoRelationship", () => {
    const model = new DefaultTrustModel();
    const from = testActor("Alice", 1);
    const to = testActor("Bob", 2);

    const metrics = model.between(from, to);
    expect(metrics.overall.value).toBe(0.0);
    expect(metrics.confidence.value).toBe(0.0);
  });

  it("betweenAsymmetric", () => {
    const model = new DefaultTrustModel();
    const alice = testActor("Alice", 1);
    const bob = testActor("Bob", 2);

    // Alice trusts Bob
    const ev = testTrustEvent(bob.id, 0.0, 0.05);
    model.updateBetween(alice, bob, ev);

    // Alice→Bob has trust
    const ab = model.between(alice, bob);
    expect(ab.overall.value).toBeGreaterThan(0.0);

    // Bob→Alice has no trust (asymmetric)
    const ba = model.between(bob, alice);
    expect(ba.overall.value).toBe(0.0);
  });

  it("evidenceCappedAt100", () => {
    const model = new DefaultTrustModel();
    const a = testActor("Alice", 1);

    // Add 150 updates
    for (let i = 0; i < 150; i++) {
      const ev = testTrustEvent(a.id, 0.0, 0.01);
      model.update(a, ev);
    }

    const metrics = model.score(a);
    expect(metrics.evidence.length).toBeLessThanOrEqual(100);
  });

  it("decayDirectedTrust", () => {
    const model = new DefaultTrustModel();
    const from = testActor("Alice", 1);
    const to = testActor("Bob", 2);

    // Build directed trust
    for (let i = 0; i < 5; i++) {
      const ev = testTrustEvent(to.id, 0.0, 0.1);
      model.updateBetween(from, to, ev);
    }

    const before = model.between(from, to);

    // Decay from actor's perspective
    model.decay(from, 10 * 86400);

    const after = model.between(from, to);
    expect(after.overall.value).toBeLessThan(before.overall.value);
  });
});

File: rust/tests/trust_test.rs
use std::collections::BTreeMap;
use serde_json::Value;

use eventgraph::actor::{Actor, ActorStatus, ActorType};
use eventgraph::event::{create_event, NoopSigner};
use eventgraph::trust::{DefaultTrustModel, TrustConfig, TrustModel};
use eventgraph::types::*;

// ── Helpers ─────────────────────────────────────────────────────────────

fn test_actor(name: &str) -> Actor {
    let id = ActorId::new(name).unwrap();
    let pk = PublicKey::new([0u8; 32]);
    Actor::new(
        id,
        pk,
        name.to_string(),
        ActorType::AI,
        BTreeMap::new(),
        1_000_000_000,
        ActorStatus::Active,
    )
}

fn test_actor_with_key(name: &str, key_byte: u8) -> Actor {
    let id = ActorId::new(name).unwrap();
    let pk = PublicKey::new([key_byte; 32]);
    Actor::new(
        id,
        pk,
        name.to_string(),
        ActorType::AI,
        BTreeMap::new(),
        1_000_000_000,
        ActorStatus::Active,
    )
}

fn make_evidence(content: BTreeMap<String, Value>) -> eventgraph::event::Event {
    let source = ActorId::new("system").unwrap();
    let event_type = EventType::new("trust.updated").unwrap();
    let cause_id = eventgraph::event::new_event_id();
    let conv_id = ConversationId::new("conv_test").unwrap();
    let prev_hash = Hash::zero();
    let signer = NoopSigner;

    create_event(
        event_type,
        source,
        content,
        vec![cause_id],
        conv_id,
        prev_hash,
        &signer,
        1,
    )
}

fn evidence_with_current(current: f64) -> eventgraph::event::Event {
    let mut content = BTreeMap::new();
    content.insert("current".to_string(), Value::from(current));
    make_evidence(content)
}

fn plain_evidence() -> eventgraph::event::Event {
    make_evidence(BTreeMap::new())
}

// ── Tests ───────────────────────────────────────────────────────────────

#[test]
fn test_initial_score() {
    let model = DefaultTrustModel::new(TrustConfig::default());
    let actor = test_actor("alice");
    let metrics = model.score(&actor).unwrap();

    assert!((metrics.overall.value() - 0.0).abs() < f64::EPSILON);
    assert!((metrics.confidence.value() - 0.0).abs() < f64::EPSILON);
    assert!((metrics.trend.value() - 0.0).abs() < f64::EPSILON);
    assert!(metrics.evidence.is_empty());
}

#[test]
fn test_update_increases_trust() {
    let mut model = DefaultTrustModel::new(TrustConfig::default());
    let actor = test_actor("alice");

    // current=0.5 when score is 0.0 gives delta=0.5, clamped to max_adjustment=0.1
    let ev = evidence_with_current(0.5);
    let metrics = model.update(&actor, &ev).unwrap();

    assert!((metrics.overall.value() - 0.1).abs() < 1e-9);
    assert_eq!(metrics.evidence.len(), 1);
}

#[test]
fn test_update_decreases_trust() {
    let config = TrustConfig {
        initial_trust: Score::new(0.5).unwrap(),
        ..TrustConfig::default()
    };
    let mut model = DefaultTrustModel::new(config);
    let actor = test_actor("alice");

    // current=0.2 when score is 0.5 gives delta=-0.3, clamped to -0.1
    let ev = evidence_with_current(0.2);
    let metrics = model.update(&actor, &ev).unwrap();

    assert!((metrics.overall.value() - 0.4).abs() < 1e-9);
}

#[test]
fn test_update_clamped_to_max_adjustment() {
    let mut model = DefaultTrustModel::new(TrustConfig::default());
    let actor = test_actor("alice");

    // current=1.0 when score is 0.0 gives delta=1.0, clamped to 0.1
    let ev = evidence_with_current(1.0);
    let metrics = model.update(&actor, &ev).unwrap();

    assert!((metrics.overall.value() - 0.1).abs() < 1e-9);
}

#[test]
fn test_update_deduplication() {
    let mut model = DefaultTrustModel::new(TrustConfig::default());
    let actor = test_actor("alice");

    let ev = evidence_with_current(0.5);
    let m1 = model.update(&actor, &ev).unwrap();
    // Same event again should not change anything.
    let m2 = model.update(&actor, &ev).unwrap();

    assert!((m1.overall.value() - m2.overall.value()).abs() < f64::EPSILON);
    assert_eq!(m1.evidence.len(), m2.evidence.len());
}

#[test]
fn test_update_trend_positive() {
    let mut model = DefaultTrustModel::new(TrustConfig::default());
    let actor = test_actor("alice");

    let ev = evidence_with_current(0.5);
    let metrics = model.update(&actor, &ev).unwrap();

    assert!((metrics.trend.value() - 0.1).abs() < 1e-9);
}

#[test]
fn test_update_trend_negative() {
    let config = TrustConfig {
        initial_trust: Score::new(0.5).unwrap(),
        ..TrustConfig::default()
    };
    let mut model = DefaultTrustModel::new(config);
    let actor = test_actor("alice");

    let ev = evidence_with_current(0.2);
    let metrics = model.update(&actor, &ev).unwrap();

    assert!((metrics.trend.value() - (-0.1)).abs() < 1e-9);
}

#[test]
fn test_score_in_domain() {
    let mut model = DefaultTrustModel::new(TrustConfig::default());
    let actor = test_actor("alice");

    // First update to create state.
    let ev = plain_evidence();
    model.update(&actor, &ev).unwrap();

    let domain = DomainScope::new("code.review").unwrap();
    let metrics = model.score_in_domain(&actor, &domain).unwrap();

    // No domain-specific score, so falls back to overall with halved confidence.
    assert!((metrics.overall.value() - 0.01).abs() < 1e-9); // observed_event_delta
}

#[test]
fn test_score_in_domain_fallback() {
    let mut model = DefaultTrustModel::new(TrustConfig::default());
    let actor = test_actor("alice");

    // Create enough evidence for measurable confidence.
    for _ in 0..10 {
        let ev = plain_evidence();
        model.update(&actor, &ev).unwrap();
    }

    let domain = DomainScope::new("code.review").unwrap();
    let metrics_overall = model.score(&actor).unwrap();
    let metrics_domain = model.score_in_domain(&actor, &domain).unwrap();

    // Fallback: same overall score, halved confidence.
    assert!((metrics_domain.overall.value() - metrics_overall.overall.value()).abs() < 1e-9);
    assert!((metrics_domain.confidence.value() - metrics_overall.confidence.value() / 2.0).abs() < 1e-9);
}

#[test]
fn test_decay() {
    let config = TrustConfig {
        initial_trust: Score::new(0.5).unwrap(),
        decay_rate: Score::new(0.1).unwrap(),
        ..TrustConfig::default()
    };
    let mut model = DefaultTrustModel::new(config);
    let actor = test_actor("alice");

    // Force state creation.
    let ev = plain_evidence();
    model.update(&actor, &ev).unwrap();

    // Before decay, score should be around 0.51 (0.5 + 0.01 observed_event_delta).
    let before = model.score(&actor).unwrap();

    // Decay for 1 day (86400 seconds).
    let after = model.decay(&actor, 86400.0).unwrap();

    // decay = 0.1 * 1 day = 0.1
    assert!(after.overall.value() < before.overall.value());
    let expected = before.overall.value() - 0.1;
    assert!((after.overall.value() - expected).abs() < 1e-9);
}

#[test]
fn test_decay_negative_duration() {
    let config = TrustConfig {
        initial_trust: Score::new(0.5).unwrap(),
        ..TrustConfig::default()
    };
    let mut model = DefaultTrustModel::new(config);
    let actor = test_actor("alice");

    // Negative duration should return current score unchanged.
    let metrics = model.decay(&actor, -100.0).unwrap();
    assert!((metrics.overall.value() - 0.5).abs() < 1e-9);
}

#[test]
fn test_update_between() {
    let mut model = DefaultTrustModel::new(TrustConfig::default());
    let alice = test_actor_with_key("alice", 1);
    let bob = test_actor_with_key("bob", 2);

    let ev = evidence_with_current(0.5);
    let metrics = model.update_between(&alice, &bob, &ev).unwrap();

    assert!((metrics.overall.value() - 0.1).abs() < 1e-9);
    assert_eq!(metrics.actor, *bob.id());
}

#[test]
fn test_between_no_relationship() {
    let model = DefaultTrustModel::new(TrustConfig::default());
    let alice = test_actor_with_key("alice", 1);
    let bob = test_actor_with_key("bob", 2);

    let metrics = model.between(&alice, &bob).unwrap();

    assert!((metrics.overall.value() - 0.0).abs() < f64::EPSILON);
    assert!(metrics.evidence.is_empty());
}

#[test]
fn test_between_asymmetric() {
    let mut model = DefaultTrustModel::new(TrustConfig::default());
    let alice = test_actor_with_key("alice", 1);
    let bob = test_actor_with_key("bob", 2);

    // Alice trusts Bob.
    let ev = evidence_with_current(0.8);
    model.update_between(&alice, &bob, &ev).unwrap();

    // Alice -> Bob should have trust.
    let a_to_b = model.between(&alice, &bob).unwrap();
    assert!(a_to_b.overall.value() > 0.0);

    // Bob -> Alice should be initial (0.0), asymmetric.
    let b_to_a = model.between(&bob, &alice).unwrap();
    assert!((b_to_a.overall.value() - 0.0).abs() < f64::EPSILON);
}

#[test]
fn test_evidence_capped_at_100() {
    let mut model = DefaultTrustModel::new(TrustConfig::default());
    let actor = test_actor("alice");

    for _ in 0..110 {
        let ev = plain_evidence();
        model.update(&actor, &ev).unwrap();
    }

    let metrics = model.score(&actor).unwrap();
    assert_eq!(metrics.evidence.len(), 100);
}

#[test]
fn test_decay_directed_trust() {
    let config = TrustConfig {
        initial_trust: Score::new(0.0).unwrap(),
        decay_rate: Score::new(0.1).unwrap(),
        ..TrustConfig::default()
    };
    let mut model = DefaultTrustModel::new(config);
    let alice = test_actor_with_key("alice", 1);
    let bob = test_actor_with_key("bob", 2);

    // Build up directed trust.
    let ev = evidence_with_current(0.5);
    model.update_between(&alice, &bob, &ev).unwrap();

    let before = model.between(&alice, &bob).unwrap();

    // Decay bob's trust for 1 day — should also affect directed trust targeting bob.
    model.decay(&bob, 86400.0).unwrap();

    let after = model.between(&alice, &bob).unwrap();
    assert!(after.overall.value() < before.overall.value());
}
